File: DirectedCycle.java
import java.util.HashMap;
import java.util.Map;

public class DirectedCycle {

    private enum Mark {
        WHITE, GRAY, BLACK
    };

    private Digraph g;
    private Map<String, Mark> marked;
    private boolean hasCycle;

    public DirectedCycle(Digraph g) {
        this.g = g;
        marked = new HashMap<>();
        hasCycle = false;
        for (String v : g.getVerts())
            marked.put(v, Mark.WHITE);
        for (String v : g.getVerts()) {
            if (marked.get(v) == Mark.WHITE) {
                System.out.println("Testando " + v);
                hasCycle = containsCycle(v);
                if (hasCycle)
                    break;
            }
        }
    }

    public boolean containsCycle() {
        return hasCycle;
    }

    private boolean containsCycle(String v) {
        System.out.println("estou em " + v);
        marked.put(v, Mark.GRAY);
        for (String w : g.getAdj(v)) {
            String e = v + "-" + w;
            System.out.println("e: " + e);
            if (marked.get(w) == Mark.GRAY)
                return true;
            else if (marked.get(w) == Mark.WHITE) {
                if (containsCycle(w))
                    return true;
            }
        }
        marked.put(v, Mark.BLACK);
        return false;
    }
}

File: directedcycle.cpp
#include "directedcycle.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <set>

using namespace std;

DirectedCycle::DirectedCycle(Digraph &g)
{
	hasCycle = false;
	for (auto const &w : g.getVerts())
		marked[w] = WHITE;

	for (auto const &w : g.getVerts())
	{
		if (marked[w] == WHITE)
		{
			hasCycle = containsCycle(g, w);
			if (hasCycle)
				break;
		}
	}
}

bool DirectedCycle::containsCycle()
{
	return hasCycle;
}

bool DirectedCycle::containsCycle(Digraph &g, string v)
{
	marked[v] = GRAY;
	for (auto const &w : g.getAdj(v))
	{
		string e = v + "-" + w;
		if (marked[w] == GRAY)
			return true;
		else if (marked[w] == WHITE)
		{
			bool res = containsCycle(g, w);
			if (res)
				return res;
		}
	}
	marked[v] = BLACK;
	return false;
}
